zerver/views/thumbnail.py

```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+from django.shortcuts import redirect
+from django.utils.translation import ugettext as _
+from django.http import HttpRequest, HttpResponse, HttpResponseForbidden
+from django.conf import settings
+from six.moves import urllib
+from zerver.models import UserProfile, validate_attachment_request
+from zerver.lib.request import has_request_variables
+from zerver.lib.thumbnail import generate_thumbnail_url
+
+def validate_thumbnail_request(user_profile, path):
+    # type: (UserProfile, str) -> bool
+    if path.startswith('user_uploads/'):
+        path_id = path[len('user_uploads/'):]
+        return validate_attachment_request(user_profile, path_id)
+
+    # This is an external link and we don't enforce restricted view policy here.
+    return True
+
+@has_request_variables
+def backend_serve_thumbnail(request, user_profile, url):
+    # type: (HttpRequest, UserProfile, str) -> HttpResponse
+    url = urllib.parse.unquote(url)
+    if validate_thumbnail_request(user_profile, url):
+        thumbnail_url = generate_thumbnail_url(
+            url.rsplit('?', 1)[0],
```

Comment:
This is much cleaner to do with `urllib.parse`  and the `query` parameter.

Reply:
actually I realised that the query won't even be there in the url at this point!